fuel-derive/src/compressed.rs

```diff
@@ -0,0 +1,430 @@
+        }
+    }
+
+    let def = match &s.ast().data {
+        syn::Data::Struct(v) => {
+            let variant: &synstructure::VariantInfo = &s.variants()[0];
+            let defs = field_defs(variant.ast().fields);
+            let semi = match v.fields {
+                syn::Fields::Named(_) => quote! {},
+                syn::Fields::Unnamed(_) => quote! {;},
+                syn::Fields::Unit => quote! {;},
+            };
+            quote! {
+                #[derive(Clone, serde::Serialize, serde::Deserialize)]
+                #[doc = concat!("Compresseded version of `", stringify!(#name), "`.")]
+                pub struct #compressed_name #g #w_structure #defs #semi
+            }
+        }
+        syn::Data::Enum(_) => {
+            let variant_defs: TokenStream2 = s
+                .variants()
+                .iter()
+                .map(|variant| {
+                    let vname = variant.ast().ident.clone();
+                    let defs = field_defs(variant.ast().fields);
+                    quote! {
+                        #vname #defs,
+                    }
+                })
+                .collect();
+
+            quote! {
+                #[derive(Clone, serde::Serialize, serde::Deserialize)]
+                #[doc = concat!("Compresseded version of `", stringify!(#name), "`.")]
+                pub enum #compressed_name #g #w_structure { #variant_defs }
+            }
+        }
+        syn::Data::Union(_) => panic!("unions are not supported"),
+    };
+
+    let compress_per_variant = s.each_variant(|variant| {
+        let vname = variant.ast().ident.clone();
+        let construct = match &s.ast().data {
+            syn::Data::Struct(_) => quote! { #compressed_name },
+            syn::Data::Enum(_) => quote! {#compressed_name :: #vname },
+            syn::Data::Union(_) => unreachable!(),
+        };
+        construct_compressed(&construct, variant)
+    });
+
+    let decompress_per_variant =
+        each_variant_compressed(&s, &quote! {#compressed_name}, |variant| {
+            let vname = variant.ast().ident.clone();
+            let construct = match &s.ast().data {
+                syn::Data::Struct(_) => quote! { #name },
+                syn::Data::Enum(_) => quote! {#name :: #vname },
+                syn::Data::Union(_) => unreachable!(),
+            };
+            construct_decompress(&construct, variant)
+        });
+
+    let impls = s.gen_impl(quote! {
+        gen impl ::fuel_compression::Compressible for @Self #w_impl {
+            type Compressed = #compressed_name #g;
+        }
+
+        gen impl<Ctx, E> ::fuel_compression::CompressibleBy<Ctx, E> for @Self #w_impl_field_bounds_compress {
+            async fn compress(&self, ctx: &mut Ctx) -> Result<Self::Compressed, E> {
+                Ok(match self { #compress_per_variant })
+            }
+        }
+        gen impl<Ctx, E> ::fuel_compression::DecompressibleBy<Ctx, E> for @Self #w_impl_field_bounds_decompress {
```

Comment:
We need to have a separate derive macro for the `DecompressibleBy` because not all types that derive `CompressibleBy` want to derive `DecompressibleBy`. The Coin or Message is an example. They don't have some fields and during decompressing, we need to fetch them from database

Reply:
Split the macro in b8011ce3